spdx-utils/src/main/kotlin/SpdxExpression.kt

```diff
@@ -123,6 +123,17 @@ sealed class SpdxExpression {
      */
     abstract fun validate(strictness: Strictness)
 
+    /**
+     * Return all valid license choices for this SPDX expression, by converting it to the
+     * [disjunctive normal form][disjunctiveNormalForm] and collecting all disjunct expressions.
+     */
+    fun validChoices(): Set<SpdxExpression> = this.disjunctiveNormalForm().validChoicesDnf()
+
+    /**
+     * Internal implementation of [validChoices], assuming that this expression is already in disjunctive normal form.
+     */
+    protected open fun validChoicesDnf(): Set<SpdxExpression> = setOf(this)
```

Comment:
Why are you adding this `internal` function? ..e.g. why not going for a single `validChoices` function?

Reply:
Because it was easier to implement it this way. The algorithm requires that the expression is in DNF, and it was difficult to implement `validChoices` for `SdpxCompoundExpression` when it is not clear if the expression is already in DNF, especially because the function is recursive.

Reply:
Ah ok - makes sense, I was asking as I thought the below `when` cases could be made redundant and thus the code simplified - but that'd add complexity in other places then.